service/src/main/java/com/epam/esm/servise/impl/CertificateServiceImpl.java

```diff
@@ -0,0 +1,171 @@
+            }
+            repository.addEntity(certificate);
+            saveCertificatesTag();
+        } else {
+            throw new IncorrectDataException(languageMassage.getMessage("message.not.valid"));
+        }
+    }
+
+    @Override
+    @Transactional
+    public void updateEntity(long id, OnlyCertificate onlyCertificate) {
+        Optional<Certificate> c = repository.getEntityById(id);
+        if (c.isPresent()) {
+            onlyCertificate.setId(id);
+            if (onlyCertificate.getCertificateName() == null)
+                onlyCertificate.setCertificateName(c.get().getCertificateName());
+            if (onlyCertificate.getDescription() == null)
+                onlyCertificate.setDescription(c.get().getDescription());
+            if (onlyCertificate.getPrice() <= 0)
+                onlyCertificate.setPrice(c.get().getPrice());
+            if (onlyCertificate.getDuration() <= 0) {
+                onlyCertificate.setDuration(c.get().getDuration());
+                onlyCertificate.setLastUpdateDate(LocalDateTime.now().plusDays(c.get().getDuration()));
+            } else {
+                onlyCertificate.setLastUpdateDate(LocalDateTime.now().plusDays(onlyCertificate.getDuration()));
+            }
+            if (onlyCertificate.getCreateDate() == null)
+                onlyCertificate.setCreateDate(c.get().getCreateDate());
+            if (onlyCertificate.getLastUpdateDate() == null)
+                onlyCertificate.setLastUpdateDate(c.get().getLastUpdateDate());
+            Certificate certificate = certificateMapper.mapFrom(onlyCertificate);
+
+
+            if (certificateValidator.isValid(certificate)) {
+                repository.updateEntity(certificate);
+                saveCertificatesTag();
+            } else {
+                throw new IncorrectDataException(languageMassage.getMessage("message.not.valid"));
+            }
+        } else {
+            throw new NoSuchEntityException(languageMassage.getMessage("message.certificate.with.id"));
+        }
+    }
+
+    @Override
+    @Transactional
+    public Optional<ModelCertificate> getEntity(long id) {
+        Optional<Certificate> c = repository.getEntityById(id);
+        if (c.isEmpty()) {
+            throw new NoSuchEntityException(languageMassage.getMessage("message.certificate.with.id"));
+        }
+        return c.map(readMapper::mapFrom);
+    }
+
+    @Override
+    @Transactional
+    public void deleteEntity(long id) {
+        Optional<Certificate> c = repository.getEntityById(id);
+        if (c.isPresent()) {
+            repository.deleteEntity(c.get());
+        } else {
+            throw new NoSuchEntityException(languageMassage.getMessage("message.certificate.with.id"));
+        }
+    }
+
+    @Override
+    @Transactional
+    public int countAllCertificates() {
+        return repository.countAllCertificates();
+    }
+
+
+    @Transactional
+    public void saveCertificatesTag() {
```

Comment:
Useless method 😀

Line 127: get all tags
Line 129: get certificates by names where you pass tag name as input.
Result: always empty list

Reply:
According to my business logic, when adding a certificate or a tag verifies and connects certificates and tags.  Wrong my business logic!